src/Pos/Generator/BlockEvent.hs

```diff
@@ -18,49 +19,201 @@ module Pos.Generator.BlockEvent
+buildBlockchainForest :: a -> Map Path a -> BlockchainForest a
+buildBlockchainForest defElem elements =
+    fmap (buildBlockchainTree defElem) . Map.fromListWith Map.union $ do
+        (Path path, a) <- Map.toList elements
+        case Seq.viewl path of
+            Seq.EmptyL -> error
+                "buildBlockchainForest: precondition violated, empty path"
+            pathSegment Seq.:< path' ->
+                [(pathSegment, Map.singleton (Path path') a)]
+
+buildBlockchainTree :: a -> Map Path a -> BlockchainTree a
+buildBlockchainTree defElem elements =
+    let
+        topPath = Path Seq.empty
+        topElement = Map.findWithDefault defElem topPath elements
+        -- 'otherElements' has its empty path deleted (if there was one in the
+        -- first place), so it satisfies the precondition of 'buildBlockchainForest'
+        otherElements = Map.delete topPath elements
+    in
+        BlockchainTree topElement (buildBlockchainForest defElem otherElements)
+
+-- Inverse to 'buildBlockchainForest'.
+flattenBlockchainForest' :: BlockchainForest a -> Map Path a
+flattenBlockchainForest' =
+    Map.fromList . flattenBlockchainForest (Path Seq.empty)
+
+flattenBlockchainForest :: Path -> BlockchainForest a -> [(Path, a)]
+flattenBlockchainForest (Path prePath) forest = do
+    (pathSegment, subtree) <- Map.toList forest
+    flattenBlockchainTree (Path $ prePath Seq.|> pathSegment) subtree
+
+flattenBlockchainTree :: Path -> BlockchainTree a -> [(Path, a)]
+flattenBlockchainTree prePath tree = do
+    let BlockchainTree a forest = tree
+    (prePath, a) : flattenBlockchainForest prePath forest
+
+type MonadGenBlockchain ctx m =
+    ( MonadBlockGen ctx m
+    , MonadThrow m )
+
+genBlocksInForest ::
+       (MonadGenBlockchain ctx m, RandomGen g)
+    => AllSecrets
+    -> BlockchainForest BlockDesc
+    -> RandT g m (BlockchainForest BlundDefault)
+genBlocksInForest secrets =
+    traverse $ mapRandT withClonedGState . genBlocksInTree secrets
+
+genBlocksInTree ::
+       (MonadGenBlockchain ctx m, RandomGen g)
+    => AllSecrets
+    -> BlockchainTree BlockDesc
+    -> RandT g m (BlockchainTree BlundDefault)
+genBlocksInTree secrets blockchainTree = do
+    let
+        BlockchainTree blockDesc blockchainForest = blockchainTree
+        txGenParams = case blockDesc of
+            BlockDescDefault  -> def
+            BlockDescCustom p -> p
+        blockGenParams = BlockGenParams
+            { _bgpSecrets     = secrets
+            , _bgpBlockCount  = 1
+            , _bgpTxGenParams = txGenParams
+            , _bgpInplaceDB   = True
+            }
+    -- Partial pattern-matching is safe because we specify
+    -- blockCount = 1 in the generation parameters.
+    OldestFirst [block] <- genBlocks blockGenParams
+    forestBlocks <- genBlocksInForest secrets blockchainForest
+    return $ BlockchainTree block forestBlocks
+
+-- Precondition: paths in the structure are non-empty.
+genBlocksInStructure ::
```

Comment:
Alright, let's look at it step-by-step.

1. The lookup function for annotations:
    ```
    getAnnotation path =
                Map.findWithDefault BlockDescDefault path annotations
    ```
    If the user had some preferences for how the block at `path` should be generated (there's an annotation), we'll return it. Otherwise, we return `BlockDescDefault` (the user doesn't care how the particular block is generated).

2. Collect all paths with annotations.
   ```        paths = foldMapOf folded
                  (\path -> Map.singleton path (getAnnotation path))
                  s
   ```
   Here we traverse the structure the first time, finding all paths and desired generation strategies (`BlockDesc`) for these paths. The result is `Map Path BlockDesc` — a set of paths, and a block description for each.

3. Assemble the paths into a tree.
    ```
    descForest = buildBlockchainForest BlockDescDefault paths
    ```
    The result is a tree where we have a node for each path, and the nodes contain `BlockDesc` associated with these paths. One case to be handled is when we have, say, a path like `kek/mems`, but not just `kek`. In this case, what do we put into the node at path `kek`? This is why we have defaulting again, passing `BlockDescDefault` to fill these holes. By construction, if we later use one of the paths in `paths` to find a node in `descForest`, the lookup will succeed.

5. Generate blocks in the tree.
    ```
    blockForest <- genBlocksInForest secrets descForest
    ```
    The `genBlocksInForest` function traverses the tree and generates a block for each block description. 

6. A function to get the block in the tree. Now that we have a tree with the actual blocks, we flatten it back to get the map from paths to blocks. (Before this, we had a map from paths to block descriptions). By construction, this map contains all paths we had before, so it's just a lookup.
    ```
            getBlock path = Map.findWithDefault
                (error "genBlocksInStructure: impossible happened")
                path
                (flattenBlockchainForest' blockForest)
    ```
7. Replace the paths with actual blocks in the original structure:
    ```
    fmap getBlock s
    ```

Hope this helps.